feat: persist expose session

ktaube opened the pull request:
Save the expose session to `{cwd}/.robocorp/expose_session.json`.

## Preview


https://github.com/robocorp/robo/assets/1479451/0954cbbb-af5a-4b35-8085-7ac8e840475f

Commit:
feat: persist expose session in .robocorp/expose_session.json

Commit:
feat: refactor expose session handling and read from JSON file

Commit:
feat: refactor expose session JSON writing

Commit:
feat: add confirmation

Commit:
feat: add --expose-y

Commit:
fix: black

Commit:
fix: specs

fabioz commented on action_server/src/robocorp/action_server/_server_expose.py:
```diff
@@ -27,6 +28,42 @@ class BodyPayload(BaseModel):
+
+
+def get_expose_session_path() -> str:
+    return os.path.join(os.getcwd(), ".robocorp", "expose_session.json")
```
Humm, instead of persisting in the current dir, it could be persisted in the settings dir, since that dir is already tied to the current directory (i.e.: `get_settings().datadir`), that way the user doesn't get its current directory dirty due to that.

ktaube replied:
Moved the `expose_session.json` to `settings.datadir`.
Using `get_settings` in expose loop threw "not initialized" error though, so I just changed it to pass `datadir` in process args.

fabioz commented on action_server/src/robocorp/action_server/_server_expose.py:
```diff
@@ -27,6 +28,42 @@ class BodyPayload(BaseModel):
+        log.debug(f"🗂️ Reading expose_session.json path={expose_session_path}")
+        with open(expose_session_path, "r") as f:
+            session_json = ExposeSessionJson(**json.load(f))
+    except FileNotFoundError:
```
I think it should also handle json loading errors and issues creating the ExposeSessionJson constructor (i.e.: in case the format changes in the future we don't want errors here in case the user for some reason uses 2 different versions in the same directory).

kariharju replied:
This is actually a really good point. 
Having some way to handle possible changes in the JSON as we can expect things to develop and change at a rapid rate.

ktaube replied:
Added `ValidationError` handler that logs the error.
We're using pydantic model for validating the input, so should be easy to handle changes. We can just set proper defaults for updated fields or implement a new schema for new data shape.

fabioz replied:
@ktaube I think it should also handle JSONDecodeError (i.e.: if by any reason the file becomes corrupted I believe it can be silently ignored instead of showing an exception to the user).

kariharju commented on action_server/src/robocorp/action_server/cli.py:
```diff
@@ -114,10 +115,10 @@ def _create_parser():
-        dest="expose_session",
-        help="Restart action server with existing expose session",
-        default=None,
+        "--expose-y",
```
Naming is hard... could this be:
`--expose-allow-reuse` or something.

ktaube replied:
Renamed to `--expose-allow-reuse` 👍

Commit:
feat: use settings.datadir to store expose session

settings were not initiaized in expose loop so I'm passing the datadir
as argument

Commit:
chore: rename expose-y to expose-allow-reuese

Commit:
feat: handle ValidationError

Commit:
fix: formatting and typings

fabioz commented on action_server/src/robocorp/action_server/_server_expose.py:
```diff
@@ -27,6 +28,41 @@ class BodyPayload(BaseModel):
+        pass
+    except ValidationError as e:
+        log.error("Failed to load previous expose session", e)
+        pass
```
Minor: as you already have a statement here the `pass` is redundant.

Commit:
fix: error handling in reading expose session JSON

fabioz reviewed (approved):
LGTM!
